Feature: Add Silent Payment sending support

MorenoProg opened the pull request:
This PR introduces support for sending silent payments (BIP-0352), integrated into the desktop Qt interface.

**Key Notes**

- Silent payments can be sent from standard wallets with BIP32 keystores only. This ensures all UTXOs in the wallet are eligible for the shared secret derivation.  Legacy or imported wallets may include uncompressed keys (for P2PKH), which are not compatible.
- Invoices involving silent payments are not persisted, as Electrum’s current system tracks invoice status via the scriptpubkeys of the invoice outputs. This approach doesn’t work for silent payments, since their final scriptpubkeys are only known at the point of transaction creation, not at invoice creation. Supporting this would have required intrusive changes to Electrum’s code base, with little practical benefit.
- ~RBF is disabled for silent payment transactions to prevent potential fund loss. If a user shares the same seed across two devices, broadcasts the transaction from one, and later bumps the fee on the other with added or changed inputs, the shared secret used in silent payments is broken, causing the recipient to miss the payment. If a transaction gets stuck in the mempool, users can still use CPFP as an alternative.~
RBF is allowed for silent payment transactions only when the fee bump is performed by the same wallet instance that created the transaction.

Commit:
adding silent payment core logic

Commit:
adding wallet and transaction integration

Commit:
adding persistence, avoiding persistence for silent payment invoices

Commit:
adding guards in sign_transaction

Commit:
adding silent payment support to payment identifier

Commit:
adding to contacts

Commit:
gui: main_window integration

Commit:
integrate send tab, handle password, fix contacts for silent payments

Commit:
adding silent payment info to wallet_info_dialog

Commit:
integrating confirm_tx_dialog

Commit:
gui: integrating transaction dialog

# Conflicts:
#	electrum/gui/qt/transaction_dialog.py

Commit:
fix double quotes in f-string

f321x commented on electrum/gui/qt/transaction_dialog.py:
```diff
@@ -403,7 +403,7 @@ def on_context_menu_for_outputs(self, pos: QPoint):
             if self.wallet.is_mine(addr):
                 show_list += [(_("Address Details"), lambda: self.main_window.show_address(addr, parent=self))]
-            copy_list += [(_(f"Copy {"On Chain " if out.is_silent_payment() else ""}Address"),
+            copy_list += [(_(f"Copy {'On Chain ' if out.is_silent_payment() else ''}Address"),
```
`_()` translation of f-strings doesn't work, see https://github.com/spesmilo/electrum/blob/cae71222d29babec9387f7a05e261dab68376241/electrum/i18n.py#L52.
You can use `.format()` instead.

Commit:
fix translation: avoid f-string in _() call

accumulator commented:
According to updated [BIP21 specification](https://github.com/bitcoin/bips/blob/master/bip-0321.mediawiki) silent payment addresses should be put in a `sp` query attribute. I don't see that supported in this PR.

Commit:
integrate BIP21-style silent payment URI support

MorenoProg commented:
@accumulator Yes thanks! I completely missed that bip21 had specifications for bip352. I addressed this in my most recent commit. It was more complicated than I thought, especially dealing with the absence of fallback addresses in wallets that can't send silent payments.
For example 'bitcoin:?sp=validspaddr' is a valid bip32-URI and therefore a valid payment identifier imo, but it is unusable in non-silent-payment-wallets. I saw no other way but to inject wallet-silent-payment-capability into the relevant functions inside payment_identifier.py.
This could be simplified by marking the payment identifier as invalid during BIP21 URI parsing, but that would be misleading — the URI itself is valid, even if the current wallet cannot process it.

Commit:
replace SILENT_PAYMENT_DUMMY_SPK with DummyAddress.SILENT_PAYMENT

Commit:
support silent payment addresses in contacts

accumulator commented on electrum/payment_identifier.py:
```diff
@@ -463,18 +468,35 @@ async def _do_notify_merchant(
                 on_finished(self)
 
-    def get_onchain_outputs(self, amount):
+    def get_onchain_outputs(self, amount, bip21_use_fallback=False):
```
Instead of `bip21_use_fallback` I'd prefer a more explicit `allow_silent_payment' kwarg.

accumulator commented on electrum/payment_identifier.py:
```diff
@@ -569,7 +595,7 @@ def _get_error_from_invoiceerror(self, e: 'InvoiceError') -> str:
         return error
 
-    def get_fields_for_GUI(self) -> FieldsForGUI:
+    def get_fields_for_GUI(self, *, bip21_prefer_fallback=False) -> FieldsForGUI:
```
Don't parameterize `get_fields_for_GUI`. It is fine to return a silent payment address here, even if the final payment is sent to a fallback address

SomberNight commented on electrum/wallet.py:
```diff
@@ -2671,13 +2735,21 @@ def sign_transaction(
 
+        if any(o.scriptpubkey == SILENT_PAYMENT_DUMMY_SPK for o in tx.outputs()):
+            raise SilentPaymentException("tried to sign tx with silent payment dummy SPK!")
+
```
It is ok not letting the user manually create txs that pay to these DummyAddress(es).

Regarding what type of exception is raised to the user though, and from where, not sure, I haven't checked, sorry. I suppose the same issue is present for the existing DummyAddress(es)? If so, we can consider it a pre-existing issue unrelated to the PR.

accumulator replied:
We can check for `DummyAddress`es in `PaymentIdentifier` and mark as `INVALID` if found. This would let the GUI do the right thing automatically.

MorenoProg replied:
I've added guards for DummyAddress.SILENT_PAYMENT in [1fc7ee5](https://github.com/spesmilo/electrum/pull/9900/commits/1fc7ee5861d25c687400c88247a86e4c3ec5bc46). I checked the behavior for the other two Dummys: CHANNEL is handled gracefully when the user tries to sign the tx. SWAP leads to an unhandled exception after being asked to enable Nostr.

accumulator commented:
> @accumulator Yes thanks! I completely missed that bip21 had specifications for bip352. I addressed this in commit [b1d4144](https://github.com/spesmilo/electrum/pull/9900/commits/b1d41447ebf8917c2e71da2358535b01029438ed). It was more complicated than I thought, especially dealing with the absence of fallback addresses in wallets that can't send silent payments. For example 'bitcoin:?sp=validspaddr' is a valid bip32-URI and therefore a valid payment identifier imo, but it is unusable in non-silent-payment-wallets. I saw no other way but to inject wallet-silent-payment-capability into the relevant functions inside payment_identifier.py. This could be simplified by marking the payment identifier as invalid during BIP21 URI parsing, but that would be misleading — the URI itself is valid, even if the current wallet cannot process it.

Yes, we currently do the same for lightning-only payment identifiers; The PI is valid, even if the wallet does not support lightning.

This leads to silent payment specific parameters and code in quite a few places.. I've added some review comments.

Commit:
refactor payment_identifier

# Conflicts:
#	electrum/gui/qt/send_tab.py

accumulator commented on electrum/payment_identifier.py:
```diff
@@ -279,12 +283,13 @@ def parse(self, text: str):
             self.set_state(PaymentIdentifierState.AVAILABLE)
         elif self.contacts and (contact := self.contacts.by_name(text)):
-            if contact['type'] == 'address':
+            if contact['type'] in ('address', 'sp_address'):
```
Hmm.. We could re-parse here, but `contact_list.py` is not consistent here either with openalias vs lnaddress..

Ok, leave as is for now..

accumulator commented:
Have you considered OpenAlias resolving to a silent payment address? This would seem like an ideal combination..

https://github.com/MorenoProg/electrum/blob/1fc7ee5861d25c687400c88247a86e4c3ec5bc46/electrum/payment_identifier.py#L345-L358

accumulator commented:
Getting an exception when emits `remove_transaction` (e.g. replaced by fee tx)

```
Traceback (most recent call last):
  File "/home/sander/projects/electrum/electrum/electrum/util.py", line 1206, in wrapper
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/electrum/interface.py", line 644, in wrapper_func
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/electrum/interface.py", line 670, in run
    await self.open_session(ssl_context=ssl_context)
  File "/home/sander/projects/electrum/electrum/electrum/interface.py", line 840, in open_session
    async with self.taskgroup as group:
               ^^^^^^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/packages/aiorpcx/curio.py", line 304, in __aexit__
    await self.join()
  File "/home/sander/projects/electrum/electrum/electrum/util.py", line 1381, in join
    task.result()
  File "/home/sander/projects/electrum/electrum/electrum/util.py", line 1511, in run_tasks_wrapper
    await self._run_tasks(taskgroup=taskgroup)
  File "/home/sander/projects/electrum/electrum/electrum/synchronizer.py", line 79, in _run_tasks
    async with taskgroup as group:
               ^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/packages/aiorpcx/curio.py", line 304, in __aexit__
    await self.join()
  File "/home/sander/projects/electrum/electrum/electrum/util.py", line 1381, in join
    task.result()
  File "/home/sander/projects/electrum/electrum/electrum/synchronizer.py", line 203, in _on_address_status
    await self._request_missing_txs(hist)
  File "/home/sander/projects/electrum/electrum/electrum/synchronizer.py", line 222, in _request_missing_txs
    async with OldTaskGroup() as group:
               ^^^^^^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/packages/aiorpcx/curio.py", line 304, in __aexit__
    await self.join()
  File "/home/sander/projects/electrum/electrum/electrum/util.py", line 1381, in join
    task.result()
  File "/home/sander/projects/electrum/electrum/electrum/synchronizer.py", line 244, in _get_transaction
    self.adb.receive_tx_callback(tx)
  File "/home/sander/projects/electrum/electrum/electrum/util.py", line 1235, in func_wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/electrum/address_synchronizer.py", line 440, in receive_tx_callback
    self.add_transaction(tx, allow_unrelated=True)
  File "/home/sander/projects/electrum/electrum/electrum/address_synchronizer.py", line 325, in add_transaction
    self.remove_transaction(tx_hash2)
  File "/home/sander/projects/electrum/electrum/electrum/address_synchronizer.py", line 379, in remove_transaction
    self._remove_transaction(txid)
  File "/home/sander/projects/electrum/electrum/electrum/address_synchronizer.py", line 420, in _remove_transaction
    self.db.remove_silent_payment_address(txo.address)
  File "/home/sander/projects/electrum/electrum/electrum/json_db.py", line 42, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sander/projects/electrum/electrum/electrum/wallet_db.py", line 1484, in remove_silent_payment_address
    assert isinstance(onchain_addr, str)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError

```

Commit:
make pi dummy address resistent, refactoring

MorenoProg commented on electrum/payment_identifier.py:
```diff
@@ -275,12 +278,23 @@ def parse(self, text: str):
-            self.spk_is_address = is_address
+            if bitcoin.script_to_address(scriptpubkey) == DummyAddress.SILENT_PAYMENT:
+                sp_address = self.parse_address(text)
+                if sp_address == DummyAddress.SILENT_PAYMENT:
```
I added guards for manually entering DummyAddress for all types (at least for spk, bip21 and multiline). If we care we could also guard against it when creating a contact, which is not handled yet.

MorenoProg commented:
> Have you considered OpenAlias resolving to a silent payment address? This would seem like an ideal combination..
> 
> https://github.com/MorenoProg/electrum/blob/1fc7ee5861d25c687400c88247a86e4c3ec5bc46/electrum/payment_identifier.py#L345-L358

I looked into it. Do you think the silent payment address can just be placed into the `recipient_address `-field? I haven't found any specs about handling different address types (like in bip21), all I found was this on the official OpenAlias Website:

> As this is an open standard that is meant to be extensible, defining additional pairs is up to you. Your client-side application may require certain key-value pairs as a minimum, and you should make that information easily available. If you have a use-case where you feel certain key-value pairs may provide widespread use and benefit, please reach out to us and we can include it in the standard if appropriate.

If we just place the sp_address into the `recipient_address `-field it would be quite easy to implement I guess. I checked the code, and only a few places would need adjustments. That said, there's a 'TODO' note about refactoring in one of the relevant sections, so I waited with making any changes for now.

https://github.com/MorenoProg/electrum/blob/1fc7ee5861d25c687400c88247a86e4c3ec5bc46/electrum/payment_identifier.py#L687-L694

Commit:
add openalias silent payment support

MorenoProg commented:
@accumulator I added openalias recognizing sp_addresses in [0c5e83d](https://github.com/spesmilo/electrum/pull/9900/commits/0c5e83d3af7a8a69ed32ad28ab6b923886950a57). It was almost nothing to change, which makes me think I missed something. I tested it by mocking the call to the openalias-server in `resolve_openalias` in `contacts.py`:

```
    @classmethod
    async def resolve_openalias(cls, url: str) -> Dict[str, Any]:
         if url == "sp@test.com":
            await asyncio.sleep(1)
            return {
                'address': "valid_sp",
                'name': "Silent Payment Test",
                'type': 'openalias',
                'validated': True
            }
```
This worked fine, also in non-sp-wallets.

accumulator commented:
> It was almost nothing to change, which makes me think I missed something.

No, you didn't miss anything, that's sufficient.

You can even remove https://github.com/MorenoProg/electrum/blob/0c5e83d3af7a8a69ed32ad28ab6b923886950a57/electrum/payment_identifier.py#L701-L703 as that is stale code

Commit:
remove stale code

MorenoProg commented:
There are still two design points where feedback is appreciated:

1. Password prompt: Since creating a silent payment requires access to the input private keys, the user is currently prompted for the password before building the transaction. They’ll get prompted again when signing. To avoid this double prompt, we could pass the password to confirm_tx_dialog (and also to transaction_dialog for previews), so it's reused. Would that be ok?

2. Disabling export/save for PSBTs involving silent payments: If a user exports a PSBT with silent payment outputs and edits it on another device, funds could be lost. We could block PSBT export in that case (informing why it's disabled), and only allow exporting `.txn` files where the transaction is already finalized. I’m aware that BIP375 addresses the issue of PSBTs involving silent payments. However, since the BIP is still in progress (open TODOs), and users who don’t interact with PSBTs aren’t affected, I think integrating this BIP can be deferred to a later stage.

Commit:
Improve silent payment UX:

- Avoid prompting user twice for password
- sign before showing preview, because silent payment psbt's make no sense
- hide `sign` and `combine option in transaction dialog

MorenoProg commented:
@macgyver13 Thanks for Testing. I am using a slightly modified version of the test vectors, in which the mapping of sp_address -> taproot_output is preserved. Also, I cleaned the [test file](https://github.com/MorenoProg/electrum/blob/feature/sender-silent-payment/tests/bip-0352/sp_send_test_vectors.json) to remove irrelevant cases like receiving or taproot inputs.

accumulator commented on electrum/gui/qt/send_tab.py:
```diff
@@ -319,6 +328,8 @@ def make_tx(fee_policy, *, confirmed_only=False, base_tx=None):
                 send_change_to_lightning=self.config.WALLET_SEND_CHANGE_TO_LIGHTNING,
                 merge_duplicate_outputs=self.config.WALLET_MERGE_DUPLICATE_OUTPUTS,
+                password=password,
+                mind_silent_payments=True # always mind silent payments to detect sending to a previous sp-onchain addr
```
Shouldn't `mind_silent_payments` be conditional on `any(o.is_silent_payment() for o in outputs)`?

accumulator commented:
I think what's lacking still is a check to verify a `Tx` for correctness in the context of silent payments in `Abstract_Wallet.sign_transaction`. If for whatever reason a `Tx` is messing with its inputs right before signing, this has the potential to send funds into the void. Currently there is only a `SIGHASH` check.

Commit:
Encapsulate sp limitations in tx_dialog. Inform user about automatic signing

MorenoProg commented:
> 
> Yes, that would be better. The current branch immediately signs the transaction as soon as the dialog is opened, which is different behavior than normal payments, and might be confusing.

Agreed. I addressed this in [7511afe](https://github.com/spesmilo/electrum/pull/9900/commits/7511afe4ea82b10d7f86367ed55b409ec9ebdaf7). I re-added the sign button (disabled) and added a tool tip that informs the user about the automatic signing. My motivation for immediate signing is to never expose an incomplete sp-transaction to the user (not even for adding to history).


> Instead of passing `show_combine_menu` from outside `TxDialog` it's better to encapsulate it within the dialog, to avoid accidentily enabling these options when called from other locations (e.g. from the history tab)

Applied.

Commit:
add integrity check for silent payment transactions before signing

MorenoProg commented:
> I think what's lacking still is a check to verify a `Tx` for correctness in the context of silent payments in `Abstract_Wallet.sign_transaction`. If for whatever reason a `Tx` is messing with its inputs right before signing, this has the potential to send funds into the void. Currently there is only a `SIGHASH` check.

@accumulator I added an integrity check in [1eb1b8f](https://github.com/spesmilo/electrum/pull/9900/commits/1eb1b8f9a5bfb0f2ac53bb407c48a2edf2f6fdca). Right after tx-creation all the relevant silent payment elements (inputs and sp-outputs) are hashed and bound to the tx. This binding can happen only once. The integrity hash is verified against the actual transaction at the time of signing to ensure it hasn't been tampered with.

MorenoProg commented on electrum/wallet.py:
```diff
@@ -2104,6 +2126,44 @@ def distribute_amount(amount):
+                sp_output.scriptpubkey = tr_spk
+            assert all(not lst for lst in spks_by_sp_addr.values()), "Found remaining sp_spks, but no more outputs left"
+
+            tx.set_rbf(False) # Transaction should be final
```
> What exactly must remain invariant when RBF-modifying a tx paying to a silent payment address? AFAIU, roughly,
> 
> * the set of input prevouts must not change

The set of silent payment *eligible* input prevouts must not change. Ineligible inputs (like uncompressed p2pkh) can be added without breaking the shared secret.

> * the set of output addresses must not change

The set of output addresses derived from a silent payment address must not change. It is no problem to add or remove outputs that are not sending silent payments.

> So in particular, bumping the fee by decreasing a change output (if present) should be possible. but e.g. bumping the fee by adding a new input burns the coins. (ref [#8847 (comment)](https://github.com/spesmilo/electrum/issues/8847#issuecomment-2154822196)) Is this correct?

Yes. Both decreasing change and decreasing payment is safe, but adding an eligible input would ~burn the coins. Under the assumption that the fee bumping happens on the same device, there would be no concerns at all, because the sp-outputs could be recalculated easily if an eligible input gets added.

I have thought about this quite a bit and I don't see a way where RBF can be considered safe. Some ideas:
* If the tx has change that is above a safe threshold, and we can be sure that the strategy on the fee-bumping-device is decreasing change first, RBF could be set. However, Electrum currently is adding new inputs under 'BumpFeeStrategy.PRESERVE_PAYMENT' before trying to decrease change (though theres a FIXME for this).
* If the fee-bumping-device is only allowed to add inputs when the tx is sp-ineligible (e.g. contains no sp-eligible input or no taproot output), RBF could be set. However, this would prevent many transactions from using RBF, as they appear silent-payment-eligible from the outside.
* If the fee-bumping-device only adds an ineligible input for sp-eligible tx's, RBF could be set. However, the txin-type is fixed for deterministic wallets in electrum, so there will be no access to such utxo's anyway (let me know if you can think of a trick here).

That being said, any of these points would require the user to be forced to perform an update, once sending silent payment is available. Also, all of this assumes that both devices run Electrum; users using other wallet software for fee bumping aren’t protected.

Unfortunately, [BIP-352](https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki) doesn't address this multi-device-problem, and I haven’t found any other resources discussing it either.

RBF would definitely be nice, but I don't think its feasible. At least CPFP is possible. Also, this PR informs the user explicitly that fee bumping will not be possible and enough fee should be paid.

Let me know if you have other ideas.

SomberNight replied:
> If the tx has change that is above a safe threshold, and we can be sure that the strategy on the fee-bumping-device is decreasing change first, RBF could be set. However, Electrum currently is adding new inputs under 'BumpFeeStrategy.PRESERVE_PAYMENT' before trying to decrease change (though theres a FIXME for this).

https://github.com/spesmilo/electrum/blob/01694a363b3f1ae33d319de90fda8912d14948e1/electrum/wallet.py#L2294-L2305
Well, it's a bit complicated. `_bump_fee_through_coinchooser` does not necessarily add a new input, but yes, it might. Usually it just decreases the change, but it does it in a round-about way: it removes the change output(s), and then tries to tweak the tx to reach the desired fee: usually by re-adding a change output with a smaller value. (but the new change output *could* even end up using a different address) Also note that if the `multiple_change` option is enabled, it can get even messier: first all change outputs are removed, and then when re-adding change, it might decide to create fewer/more change outputs.

Also note that even `_bump_fee_through_decreasing_change` can end up changing the output set: if needed to reach the target fee level, it can decide to completely remove the change output.

-----

> Unfortunately, [BIP-352](https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki) doesn't address this multi-device-problem, and I haven’t found any other resources discussing it either.
> Let me know if you have other ideas.

Well, there is one general approach that we could use, that works both for RBF-ing silent payments and RBF-ing lightning funding transactions.

(1): So, the current approach is that we use the opt-in-RBF flag to store a single bit of information in a tx re whether it is *safe* to RBF it. If the flag is set, we allow all devices to RBF it. If it is not set, we don't allow any device to RBF it.

Note that historically, before `mempoolfullrbf=1`, this flag was even enforced by network policy. But I guess not anymore. Now we purely use it for our own wallet logic.

(2): We could extend (1), so that the original device that created the tx disregards the flag. That is, every time we create an interesting tx (e.g. one that pays to a silent payment addr), we store the txid (or similar) to a data structure in the wallet file (persisted). This could even be cleared up after the tx gets deeply mined.

When the user instructs the wallet to RBF a tx,
- we first check if the tx opted-in to RBF
  - if yes, early return: we can RBF it
- then we check if the tx was originally created by this current wallet instance, by looking it up in the persisted list
  - if no, early return: we cannot RBF it
    - question: do we allow `dscancel`? I guess we could always allow that.
- now we can apply arbitrary restrictions:
  - e.g. `dscancel` is allowed for lightning funding
  - for pay-to-silent-payment-addr, bump_fee is restricted to only touching amounts

I think `dscancel` we could allow unconditionally for every tx. (but that might be orthogonal to this PR)

Basically, if you restore from seed while there is a tx in the mempool, you lost the ability to RBF it.
You can only RBF "special" txs if they were created in the same wallet instance.

MorenoProg replied:
Ok. I implemented this in [c670dba](https://github.com/spesmilo/electrum/pull/9900/commits/c670dbace54f7a787f459c7e9988bc8c08449aee). This supports all fee-bumping methods, provided the bumping is done on the same wallet instance that created the tx.

There was no need to persist more information than this branch already did. A silent payment tx will not signal RBF externally, though internally, RBF is considered allowed if `tx.is_rbf_enabled() or tx.contains_silent_payment()`. And only the wallet instance that created the tx will know about `tx.contains_silent_payment()`.

MorenoProg replied:
Sure, but can you elaborate a bit on what you want to be tested exactly? Because the code I added to `bump_fee` doesn't touch any RBF logic (like strategy-methods); it just updates sp-addresses if necessary, otherwise it treats silent payment tx's as 'normal' tx's, and this is already covered in large in `test_wallet_vertical.py`.

Without writing redundant tests, I can only think of the following new meaningful test cases that:
* test if silent payment outputs are recalculated if the input set changes
* simulate a two device scenario and test if rbf is refused on the fee bumping device (though this seems trivial, since I'd just test if CannotBumpFee is raised with a tx that doesn't signal RBF).

SomberNight replied:
Yes, that sounds good:

> simulate a two device scenario and test if rbf is refused on the fee bumping device

1. test that the originating wallet instance *can* bump the fee
    - then still in same test, close the wallet, and load it back again from disk, and see it can still bump the fee
      (see e.g. https://github.com/spesmilo/electrum/blob/d6c300ebcc122d635fc40f987e2a4a2abfe21a2e/tests/test_wallet.py#L108-L112)
2. test that another wallet instance *cannot* bump the fee

> test if silent payment outputs are recalculated if the input set changes

3. create a tx that pays to e.g. two silent-payment-addrs and another simple address, maybe without creating a change output,  and then bump its fee by adding a new input and a change output.  the two SPAs should have the outputs recalculated but the simple address should remain untouched.

Commit:
allow RBF-ing silent payment transactions. Don't signal RBF, instead use it as internal flag. Only allow RBF if tx originated from same wallet instance.

Commit:
followup: exclude output value from silent_payment_integrity_hash. If included it may causes unnecessarily recalculating the outputs during fee bumping with strategy DECREASE_PAYMENT

Commit:
add tests for silent payment persistence and rbf behavior

Commit:
replace restore_wallet_from_text with restore_wallet_from_text__for_unittest

Commit:
add support for silent donations

MorenoProg commented:
Just in case @accumulator's comment has gone unnoticed, tagging @ecdsa and @SomberNight.

MorenoProg commented:
@ecdsa Thanks for the response.

Quite a bit of time went into this PR, not just from my side, but also from contributors who reviewed it and asked to add unit tests, probably under the assumption that this was something you were seriously considering merging. That impression didn’t come out of nowhere: in Electrum issue #9728, the discussion was framed around how this could be implemented, not whether it should be supported at all.  

Given that, it’s a bit frustrating to reach this stage and find that there isn’t alignment on the feature direction itself.

More generally, this doesn’t seem like just a question of waiting for community demand. Electrum is in a position to shape that demand. If silent payments are something you want to see adopted, supporting them here would directly contribute to that.

If the current stance is that this isn’t a direction you want to take, that’s fine. but having that clarity earlier would have avoided a fair amount of unnecessary work.

Still happy and eager to revisit this if your position changes in the future.
